libs/parallelism/execution/include/hpx/execution/algorithms/ensure_started.hpp

```diff
@@ -0,0 +1,377 @@
+                                done_error_value_visitor<R>{std::move(r)}, v);
+                        }
+                        else
+                        {
+                            continuations.emplace_back(
+                                [this,
+                                    // NOLINTNEXTLINE(bugprone-move-forwarding-reference)
+                                    r = std::move(r)]() {
+                                    std::visit(
+                                        done_error_value_visitor<R>{
+                                            std::move(r)},
+                                        v);
+                                });
+                        }
+                    }
+                }
+
+                void start() noexcept
+                {
+                    if (!start_called.exchange(true))
+                    {
+                        hpx::execution::experimental::start(os);
+                    }
+                }
+
+                friend void intrusive_ptr_add_ref(shared_state* p)
+                {
+                    ++p->reference_count;
+                }
+
+                friend void intrusive_ptr_release(shared_state* p)
+                {
+                    auto c = --p->reference_count;
+                    if (c == 0)
+                    {
+                        delete p;
+                    }
+                }
+            };
+
+            hpx::intrusive_ptr<shared_state> st;
+
+            template <typename S_>
+            ensure_started_sender(S_&& s, Allocator const& a)
+            {
+                using allocator_type = Allocator;
+                using other_allocator = typename std::allocator_traits<
+                    allocator_type>::template rebind_alloc<shared_state>;
+                using allocator_traits = std::allocator_traits<other_allocator>;
+                using unique_ptr = std::unique_ptr<shared_state,
+                    util::allocator_deleter<other_allocator>>;
+
+                other_allocator alloc(a);
+                unique_ptr p(allocator_traits::allocate(alloc, 1),
+                    hpx::util::allocator_deleter<other_allocator>{alloc});
+
+                new (p.get()) shared_state{std::forward<S_>(s)};
+                st = p.release();
+
+                // Eagerly start the work received until this point.
+                //
+                // P1897r3 says "When start is called on os2 [the operation
+                // state resulting from connecting the ensure_started_sender to
+                // a receiver], call execution::start(os [the operation state
+                // resulting from connecting S to ensure_started_receiver])",
+                // which would indicate that start should be called later.
+                // However, to fulfill the promise that ensure_started actually
+                // eagerly submits the sender S we call start already here.
+                st->start();
```

Comment:
If `st->start()` throws, then this will leak the memory `st` refers to.

Reply:
I take this back, `st` is an intrusive_ptr. All is well.

Reply:
However, since the memory is allocated using an arbitrary allocator, shouldn't it be deleted using the same allocator (i.e. shouldn't the shared_state store the allocator for later use)?

Reply:
Very good point, I completely missed that! I'll fix that.

Reply:
I think I've fixed these now (it was the same situation in `detach.hpp`).